Library view updates

tjacovich opened the pull request:
- Added `QueryView` to biblib documentation
- Updated response for biblib `DocumentView`

tjacovich commented:
Sorry about the number of commits. The example formatting for `QueryView` was quite finicky. Figure it'll be squashed in the merge, but it makes this seem like a much more complex PR than it actually turned out to be.

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +715,34 @@
+   "metadata": {},
+   "source": [
+    "### /query/&lt;library_id&gt;\n",
+    "Interact with a specific library. The `library_id` (string) is the unique identifier of the library, as specified in the `/libraries` GET response.\n",
```
"Interact with a specific library" is vague. Maybe something like "Add or delete a documents to a given library based on a supplied query" or something like that.

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +715,34 @@
+    "### /query/&lt;library_id&gt;\n",
+    "Interact with a specific library. The `library_id` (string) is the unique identifier of the library, as specified in the `/libraries` GET response.\n",
+    "\n",
+    "Functions allowed: add query to a library (POST, GET), delete records from a library (POST, GET).\n",
```
The GET only allows you to add documents, not remove them, right?

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +715,34 @@
+    "Functions allowed: add query to a library (POST, GET), delete records from a library (POST, GET).\n",
+    "\n",
+    "#### POST\n",
+    "Add or delete a document to a given library. Returns the number of documents added or removed\n",
```
Update both sentences to match this endpoint

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +715,34 @@
+    "Add or delete a document to a given library. Returns the number of documents added or removed\n",
+    "\n",
+    "Specify that the request is a POST method by adding `-X POST` to your request. You must include a payload, passed via the `-d` flag, in JSON format. Add this header to your request to indicate your payload is in the appropriate format: `-H \"Content-Type: application/json\"`. The payload must include:\n",
+    "  * bibcode: (list) list of bibcodes to be added or removed\n",
```
Leftover from copy/paste, update to match this endpoint

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +715,34 @@
+    "  * bibcode: (list) list of bibcodes to be added or removed\n",
+    "  * action: (string) 'add' or 'remove' to add or remove the given bibcodes from the specified library\n",
+    "\n",
+    "*Return data*:\n",
```
Update to match this endpoint

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +715,34 @@
+    "  * admin\n",
+    "  * write"
+   ]
+  },
```
Add a code block example here too, to demonstrate usage. Unfortunately it'll have to be a curl - I'm working on a python version but haven't made a PR yet

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +715,34 @@
+    "\n",
+    "Functions allowed: add query to a library (POST, GET), delete records from a library (POST, GET).\n",
+    "\n",
+    "#### POST\n",
```
Add a description and a code block for the GET as well

kelockhart reviewed (changes requested):
The formatting/rendering looks good! I have some content comments though - some smaller ones for the OpenAPI docs and some larger ones for the notebook.

tjacovich commented:
Did the recommended clean-up for biblib.yaml and added the lines in openapi_public.yaml. Updated Libraries_API.ipynb to actually contain examples and properly describe the new endpoint `QueryView` + modifications to `DocumentView`.

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +664,151 @@
+   "metadata": {},
+   "source": [
+    "### /query/&lt;library_id&gt;\n",
+    "Add or remove documents the library specified by `library_id` (string) is the unique identifier of the library, as specified in the `/libraries` GET response.\n",
```
The first half (before the comma) of this line needs some editing - I think two clauses got smushed together weirdly

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +664,151 @@
+    "Add or delete a query to/from a given library. Returns the number of documents added or removed along with the added bibcodes.\n",
+    "\n",
+    "Specify that the request is a POST method by adding `-X POST` to your request. You must include a payload, passed via the `-d` flag, in JSON format. Add this header to your request to indicate your payload is in the appropriate format: `-H \"Content-Type: application/json\"`. The payload must include:\n",
+    "  * query: (json) dict of query parameters as specified in 'search'\n",
```
"as specified in 'search'" - do you mean in the search notebook? If so, I'd say something more like "dict of query parameters (see the Search notebook for more details)"

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +664,151 @@
+    "*Return data*:\n",
+    "  * number_added: (integer) number of documents added (if 'add' is used)\n",
+    "  * number_removed: (integer) number of documents removed (if 'remove' is used)\n",
+    "  * bibcodes: (list) list of bibcodes added to the library\n",
```
Will this list be populated with the removed bibcodes if action=remove? If so, change "added" to "added/removed". Otherwise clarify what will be return if bibcodes are removed

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +664,151 @@
+    "#### GET\n",
+    "Add a document to a given library. Returns the number of documents added along with a list of added bibcodes\n",
+    "\n",
+    "  * Query based on `search` endpoint \n",
```
This wording is a little confusing - do you mean that query parameters are the same as those used in the search endpoint? If so, I'd say something like that, and mention the search notebook, like in my earlier comment (people aren't necessarily reading the whole notebook, they're mostly just looking for exactly what they need, so it won't seem repetitive)

kelockhart replied:
You might even want to write out here what the structure of the URL is, since it's a little different:
`/query/<library_id>?<query_parameters>`
or something like that

kelockhart commented on Libraries_API.ipynb:
```diff
@@ -713,6 +664,151 @@
+   "source": [
+    "\n",
+    "#### GET\n",
+    "Add a document to a given library. Returns the number of documents added along with a list of added bibcodes\n",
```
I'd say explicitly that this endpoint is only for adding bibcodes to a library, and that to remove bibcodes via query you need to use the POST endpoint

kelockhart reviewed (changes requested):
The notebook is looking much better! I have some small wording suggestions, but otherwise it's nearly there.

kelockhart reviewed (approved):
Looks good! Squash and merge when you're ready.
